Step4

jysohn0825 opened the pull request:
1. ResultView에서 레이스를 진행하게 하는 메소드 방식이 옳은 지에 대한 코멘트 부탁 드리겠습니다
2. 코틀린에서는 한 파일 내에 여러 클래스를 작성할 수 있는데, 현업에서는 굳이 적은 단위의 기능의 클래스들을 나누기 보다는 한 파일 안에서 여러 클래스를 관리하는 게 더 낫지 않냐라는 의문을 던지셔서 리뷰어님의 생각이 궁금합니다
3. 전략 패턴을 적용하고자 했으나... 제가 정확한 개념을 모르는 거 같아 다시 문의 드립니다. 전략 패턴 인터페이스를 상속 받은 자동차 전략 클래스의 move 안에도 Random이 있다면, 테스트 코드를 짤 수 없는 것이 동일하지 않나요...?

jysohn0825 committed:
kotlin basic grammar

jysohn0825 committed:
Add Calculator Production Code

without test code

jysohn0825 committed:
Add Test Code

add test code
optimizing production & test code

jysohn0825 committed:
Merge branch 'jysohn0825' into step2

jysohn0825 committed:
Fix Operaotr enum

- fix operaotr enum method (find method occur IllegalArguementException, add high order function)
- change validCheck method, not if branch but require method

jysohn0825 committed:
Revert "Fix Operaotr enum"

This reverts commit 7d92bc5eac560224e6b10632dba70cb33d3645a1.

jysohn0825 committed:
Merge commit 'f8041e14195d5728497f237ded9f2430e9640b96' into step2

jysohn0825 committed:
Fix Operator enum

change operaotr enum (add high order function, add Exception)
change method name
add test code

jysohn0825 committed:
Add RacingCar classes

- make domain package (Car, Race)
- make view package (InputView, ResultView)
- make RacingGame class

jysohn0825 committed:
Add Car method

add Car method also test code

jysohn0825 committed:
Add Race

add race method

jysohn0825 committed:
Add View

- add InputView
- add ResultView

jysohn0825 committed:
Make Racig Car Game

- add main method
- add Racing Game class

jysohn0825 committed:
Merge branch 'jysohn0825' into step3

jysohn0825 committed:
Refactoring Step3

- add README.md
- change method (Car, Race, *View)

jysohn0825 committed:
Revert "Refactoring Step3"

This reverts commit 69963eec575f2cd39c6f4a05ebee48fe87406d12.

jysohn0825 committed:
Merge commit 'ca817b393c1525c5d36c68888650230c6979d8a8' into step3

* commit 'ca817b393c1525c5d36c68888650230c6979d8a8':
  Step2 : 문자열 계산기 (#734)

jysohn0825 committed:
doc: Add README.md

- README.md 추가 (3단계, 4단계)

jysohn0825 committed:
refactor: Refactoring Step3 code

- change All class

jysohn0825 committed:
feat: Add name property in Car domain

- add name property and valid method
- add test code for valid method

jysohn0825 committed:
feat: Print Car name

jysohn0825 committed:
feat: Seperate Car by comma

- add generate Car method by name
- add test code

jysohn0825 committed:
refactor: Change ResultView

- seperate printResult for low coupling

jysohn0825 committed:
feat: Add winner logic

- add print winner logic
- add find winner test code
- add find winner logic

jysohn0825 committed:
refactor: Change ResultView

jysohn0825 committed:
refactor: Change ResultView

kyucumber commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -0,0 +1,44 @@
+import racingcar.domain.Car
+import racingcar.domain.Race
+
+class ResultView {
```
> ResultView에서 레이스를 진행하게 하는 메소드 방식이 옳은 지에 대한 코멘트 부탁 드리겠습니다

ResultView는 도메인 클래스 등을 받아 안에 담긴 데이터를 가지고 출력하는 역할을 수행하는 것이 자연스러워 보입니다.
지금처럼 전체적인 레이스 진행 등의 역할을 수행하는게 View에 들어가는게 좋은 형태로 보이진 않네요.

kyucumber commented on src/main/kotlin/racingcar/README.md:
```diff
@@ -0,0 +1,20 @@
+# 자동차 경주
```
> 코틀린에서는 한 파일 내에 여러 클래스를 작성할 수 있는데, 현업에서는 굳이 적은 단위의 기능의 클래스들을 나누기 보다는 한 파일 안에서 여러 클래스를 관리하는 게 더 낫지 않냐라는 의문을 던지셔서 리뷰어님의 생각이 궁금합니다

명확한 답은 없겠지만 저는 기본적으로는 한 클래스당 한 파일을 유지하는걸 원칙으로 합니다.
의미적으로 서로 밀접하게 관련되어 있고 파일 크기(코드 라인 수)가 합리적으로 유지되는 경우라면 같은 파일에 두는 걸 고민할 수 있겠네요.

kyucumber commented on src/main/kotlin/racingcar/domain/Race.kt:
```diff
@@ -0,0 +1,28 @@
+const val RANDOM_MAX_NUM = 9
+const val RANDOM_CONDITION_NUM = 4
+
+class Race(
+    val round: Int
+) {
```
> 전략 패턴을 적용하고자 했으나... 제가 정확한 개념을 모르는 거 같아 다시 문의 드립니다. 전략 패턴 인터페이스를 상속 받은 자동차 전략 클래스의 move 안에도 Random이 있다면, 테스트 코드를 짤 수 없는 것이 동일하지 않나요...?

전략 패턴은 전략을 교체할 수 있는 이점이 있습니다. 랜덤이 들어있는 전략 클래스의 구현체의 테스트가 안된다고 말씀하시는거라면 그 부분은 동일하게 테스트 할 수 없겠지만 그 구현체를 사용하는 클래스는 인터페이스를 교체해 테스트 할 수 있지 않을까요?

```kotlin
class Race(
    val round: Int
    val movingRule: MovingRule(인터페이스)
) {
    private fun moveCar(car: Car) {
        if (movingRule.isMove()) car.move()
    }
}

```

위와 같은 경우 테스트에서 MovingRule 인터페이스의 구현체를 무조건 이동하게 한다거나 하는 식으로 테스트 할 수 있습니다.

kyucumber commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    fun valid() {
+        require(name.length < MAX_CAR_NAME_LENGTH) { IllegalArgumentException("차 이름은 5자를 넘으면 안 됩니다.") }
```
require 내부에서 IllegalArgumentException을 반환합니다. 예외 타입이 아니라 메시지만 반환하도록 변경해주세요

kyucumber reviewed (changes requested):
안녕하세요 몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

jysohn0825 committed:
refactor: change method

- change ResultView for seperate Race
- add Rule interface and class
- add test code

kyucumber reviewed (approved):
피드백 잘 반영해주셨습니다.

다음 단계 진행을 위해 머지할게요
